src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -795,9 +795,15 @@ public ClusterState execute(ClusterState currentState) {
 
                         ClusterState updatedState = stateToProcess.clusterState;
 
-                        // if the new state has a smaller version, and it has the same master node, then no need to process it
-                        if (updatedState.version() < currentState.version() && Objects.equal(updatedState.nodes().masterNodeId(), currentState.nodes().masterNodeId())) {
-                            return currentState;
+                        if (currentState.nodes().masterNodeId() != null) {
+                            if (!currentState.nodes().masterNodeId().equals(updatedState.nodes().masterNodeId())) {
+                                logger.info("received a cluster state that has [{}] master node, while current cluster state has [{}] as master node, ignoring...", updatedState.nodes().masterNode(), currentState.nodes().masterNode());
```

Comment:
+1 on warn. Can also do this check outside of the cluster state update task? it's a shame to go into an update task. We will still need this check in the cluster state as we may have two masters publishing while we are in the join process.